### DISCUSSION TITLE
Consuming Paginated Response with Token

### DISCUSSION THREAD

**Discussion by whampt** [Jericho/ZoomNet#282] [Q&A] [answered]
Thank you for writing this library that makes it easier to manage server-to-server OAuth with Zoom Apps. I'm a Perl script monkey learning to consume APIs with C#. 

Can you assist with paginating through the response from zoomClient.PastWebinars.GetParticipantsAsync(webinarId)? Visual Studio says the code below cannot convert `System.Threading.Tasks.Task<ZoomNet.Models.PaginatedResponseWithToken<ZoomNet.Models.Participant>>` to `string`

```
// Zoom API using server-to-server OAuth
var zoomClientId = "your zoom client id";
var zoomClientSecret = "your zoom client secret";
var authorizationCode = "your zoom authorization code";
var redirectUri = "... the URI you have configured when setting up your OAuth app ...";
var connectionInfo = OAuthConnectionInfo.WithAuthorizationCode(zoomClientId, zoomClientSecret, authorizationCode,
    (newRefreshToken, newAccessToken) =>
    {
        Environment.SetEnvironmentVariable("ZOOM_OAUTH_REFRESHTOKEN", newRefreshToken, EnvironmentVariableTarget.User);
    },
    redirectUri);
var zoomClient = new ZoomClient(connectionInfo);

// ID of the webinar to list participants for
long webinarId = 9999999999;

// Make a request to the Zoom API to list webinar participants
var participants = zoomClient.PastWebinars.GetParticipantsAsync(webinarId);
var deserializedParticipants = JsonConvert.DeserializeObject(participants);

// Print the list of participants to the console
Console.WriteLine($"Webinar participants for webinar ID {webinarId}:");
foreach (var participant in participants)
{
    Console.WriteLine($"- {participant.Name}, {participant.Email}");
}
```

I've read the README, gone through the project code and looked through the Object/Assembly browser at the definition of PaginatedResponseWithToken. Still confused on what class to define the response as and how to keep returning the token to read the next page of responses. I'm happy to be pointed at a relevant tutorial or existing example. I've looked at a few tutorials, these seem to be on how to create an new API service instead of consuming an existing API. 

I would appreciate guidance on picking a tutorial on consuming paginated responses compatible with ZoomNet's way of doing things. 

Thank you

**Comment by whampt** [comment]
The Target Framework for this project is .net 7.0 as it is new and shiny. No real dependencies, Target Framework can be changed

**Comment by Jericho** [comment, marked as answer]
The first thing that jumps at me when I look at the code you posted is that you invoke an asynchronous method (`GetParticipantsAsync` in this case) but you do not await the result. The consequence of this is that your `participants` variable contains a non-completed task 
 rather than the data you expect and therefore it explains why you are getting an exception about converting a Threading.Task to another data type.

Furthermore, when you await an async method, it's considered best practice to configure the synchronization context by invoking `.ConfigureAwait(false)` which is missing from your code sample. Please note that this is not mandatory, but highly encouraged. 

The next thing that is important to understand is that most methods in the ZoomNet library that retrieve a "page" of data, such as GetParticipantsAsync for example, return an object of type `PaginatedResponseWithToken` and not simply an array of records. This explains why your `foreach (var participant in participants)` doesn't work. This object type contains several properties, some of the most useful (in my opinion) are:
- `NextPageToken` is the token you must use to subsequently request the next page of data
- `Records` contains the records you requested
- `MoreRecordsAvailable`: this boolean value indicates if there are more records available to be retrieved. For example, if 51 participants are associated with a webinar and you retrieved the first 50, this property will contain `true`. However, if only 49 people attended your webinar, this property will contain `false`.

Something else I noticed is that you are attempting to print the name of each participant to the console but the `Participant` model class does not have a `Name` property. The property you are looking for is called `DisplayName`.

Putting all this together, here's how you can improve the code you posted:
```csharp
// Configure the number of records you want to retrieve at a time
var recordsPerPage = 50;

// Asynchronously retrieve the first "page" of participants
var getParticipantsResult = await zoomClient.PastWebinars.GetParticipantsAsync(webinarId, recordsPerPage).ConfigureAwait(false);

// Loop through the records
foreach (var participant in getParticipantsResult.Records)
{
    Console.WriteLine($"- {participant.DisplayName}, {participant.Email}");
}

// Check if there are more records to be retrieved
if (getParticipantsResult.MoreRecordsAvailable)
{
    // Retrieve the second "page" of participants
    var paginationToken = getParticipantsResult.NextPageToken;
    getParticipantsResult = await client.PastWebinars.GetParticipantsAsync(scheduledWebinar.Id, recordsPerPage, paginationToken).ConfigureAwait(false);
    foreach (var participant in getParticipantsResult.Records)
    {
        Console.WriteLine($"- {participant.DisplayName}, {participant.Email}");
    }
}
```

Let me know if this helps.

**Reply by Jericho** [reply]
@whampt Did I answer your question? Was the code I provided useful?

**Comment by whampt** [comment]
Thank you for the detailed response, your code is very detailed & looks to be very helpful. You went far and above with your response to my question!

I have not gotten back to my side of this project yet.

**Comment by whampt** [comment]
The supplied code works perfectly! Thank you very much!
